### ISSUE TITLE
weave launch - conatiner died

### ISSUE THREAD

**Issue by rachit1arora** [weaveworks/weave#1729]
Hello We have an intermittent issue that weave launch container dies. 

Here is the detailed scenario 
We have a bunch of machines in network (eg 10) . For a day or so every things works fine . 
Many containers get scheduled on these hosts and get weave ip all the time. 

Sometime on one of the machine container start dieing when we need to apply weave ip. 
If we do a restart of this machine properly and do a weave launch and join the network back , every thing works fine again. 
But if we try to do weave launch on this machine WITHOUT restart . weave launch container dies 

[root@**** bin]# weave launch -iprange 172.16.0.0/8
Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
Unable to find image 'weaveworks/weave:1.0.1' locally
1.0.1: Pulling from docker.io/weaveworks/weave
Digest: sha256:acd3d09951947d56f6c96505b07314cf937324de283941dfb9b6087bd4e1d423
Status: Downloaded newer image for docker.io/weaveworks/weave:1.0.1
Container 68003c56a33c54f4cb4814fedefdbc877a64a43f23b0c471b16922b3e9cc6d5c died

---Tried Forceful restart of conatiner --- 
[root@**** bin]# docker start 68003c56a33c54f4cb4814fedefdbc877a64a43f23b0c471b16922b3e9cc6d5c
68003c56a33c54f4cb4814fedefdbc877a64a43f23b0c471b16922b3e9cc6d5c

--- Here are docker logs-----
[root@bi-paygobm-prod-023 bin]# docker logs -f weave
weave 2015/11/27 07:11:35.505177 Command line options: map[name:e6:04:94:89:4a:05 nickname:**** port:6783 wait:20 iface:ethwe iprange:172.16.0.0/8]
weave 2015/11/27 07:11:35.505213 Command line peers: []
weave 2015/11/27 07:11:55.555497 Unable to find interface ethwe

---------OUTPUT of weave launch in debug mode -----
````
 WEAVE_DEBUG=1 weave launch -iprange 172.16.0.0/8
+ SCRIPT_VERSION=1.0.1
+ '[' 1.0.1 = '(unreleased version)' ']'
+ IMAGE_VERSION=1.0.1
+ IMAGE_VERSION=1.0.1
+ DOCKERHUB_USER=weaveworks
+ BASE_EXEC_IMAGE=weaveworks/weaveexec
+ EXEC_IMAGE=weaveworks/weaveexec:1.0.1
++ cut -s -d: -f1
++ echo ''
+ PROXY_HOST=
+ PROXY_HOST=127.0.0.1
+ IP_REGEXP='[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}'
+ CIDR_REGEXP='[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}/[0-9]{1,2}'
+ '[' launch = run ']'
+ '[' launch '!=' --local ']'
+ exec_remote launch -iprange 172.16.0.0/8
+ docker run --rm --privileged --net=host -v /var/run/docker.sock:/var/run/docker.sock -v /proc:/hostproc -e PROCFS=/hostproc -e DOCKERHUB_USER=weaveworks -e VERSION -e WEAVE_DEBUG -e WEAVE_DOCKER_ARGS -e WEAVEDNS_DOCKER_ARGS -e WEAVEPROXY_DOCKER_ARGS -e WEAVE_PASSWORD -e WEAVE_PORT -e WEAVE_CONTAINER_NAME -e DOCKER_BRIDGE -e PROXY_HOST=127.0.0.1 -e WEAVE_CIDR=none weaveworks/weaveexec:1.0.1 --local launch -iprange 172.16.0.0/8
Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
+ SCRIPT_VERSION=1.0.1
+ [ 1.0.1 = (unreleased version) ]
+ IMAGE_VERSION=1.0.1
+ IMAGE_VERSION=1.0.1
+ DOCKERHUB_USER=weaveworks
+ BASE_EXEC_IMAGE=weaveworks/weaveexec
+ EXEC_IMAGE=weaveworks/weaveexec:1.0.1
+ PROXY_HOST=127.0.0.1
+ PROXY_HOST=127.0.0.1
+ IP_REGEXP=[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}
+ CIDR_REGEXP=[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}/[0-9]{1,2}
+ [ --local = run ]
+ [ --local != --local ]
+ shift 1
+ BASE_IMAGE=weaveworks/weave
+ BASE_DNS_IMAGE=weaveworks/weavedns
+ IMAGE=weaveworks/weave:1.0.1
+ DNS_IMAGE=weaveworks/weavedns:1.0.1
+ PROCFS=/hostproc
+ DOCKER_BRIDGE=docker0
+ CONTAINER_NAME=weave
+ DNS_CONTAINER_NAME=weavedns
+ BRIDGE=weave
+ CONTAINER_IFNAME=ethwe
+ MTU=65535
+ PORT=6783
+ HTTP_PORT=6784
+ DNS_HTTP_PORT=6785
+ DNS_HTTP_IFNAME=eth0
+ PROXY_PORT=12375
+ PROXY_CONTAINER_NAME=weaveproxy
+ id -u
+ [ 0 = 0 ]
+ uname -s -r
+ sed -n -e s|^\([^ ]*\) \([0-9][0-9]*\)\.\([0-9][0-9]*\).*|\1 \2 \3|p
+ read sys maj min
+ [ Linux != Linux ]
+ [ ( 3 -eq 3 -a 10 -ge 8 ) -o 3 -gt 3 ]
+ command_exists ip
+ command -v ip
+ ip netns list
+ command_exists nsenter
+ command -v nsenter
+ check_docker_version
+ docker -v
+ sed -n -e s|^Docker version \([0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\).*|\1|p
+ DOCKER_VERSION=1.3.1
+ [ -z 1.3.1 ]
+ [ 1.3.1 = 1.3.0 ]
+ echo 1.3.1
+ cut -d. -f 1
+ DOCKER_VERSION_MAJOR=1
+ echo 1.3.1
+ cut -d. -f 2
+ DOCKER_VERSION_MINOR=3
+ echo 1.3.1
+ cut -d. -f 3
+ DOCKER_VERSION_PATCH=1
+ [ 3 -gt 0 ]
+ COMMAND=launch
+ shift 1
+ check_not_running weave weaveworks/weave
+ docker inspect --format={{.State.Running}} {{.Config.Image}} weave
+ docker rm weave
+ create_bridge
+ [ ! -d /sys/class/net/weave ]
+ [  != --without-ethtool ]
+ ethtool -K weave tx off
+ ip link set dev weave up
+ configure_arp_cache weave
+ sh -c echo 5 >/proc/sys/net/ipv4/neigh/weave/base_reachable_time &&
              echo 2 >/proc/sys/net/ipv4/neigh/weave/delay_first_probe_time &&
              echo 1 >/proc/sys/net/ipv4/neigh/weave/ucast_solicit
+ cat /sys/class/net/weave/address
+ PEERNAME=e6:04:94:89:4a:05
+ is_cidr -iprange
+ echo -iprange
+ grep -E ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}/[0-9]{1,2}$
+ CONTAINER_PORT=6783
+ ARGS=
+ IPRANGE=
+ IPRANGE_SPECIFIED=
+ [ 2 -gt 0 ]
+ [ 2 -gt 1 ]
+ IPRANGE=172.16.0.0/8
+ IPRANGE_SPECIFIED=1
+ shift 2
+ [ 0 -gt 0 ]
+ eval set --
+ set --
+ [ -z 1 ]
+ hostname
+ docker run --privileged -d --name=weave -p 6783:6783/tcp -p 6783:6783/udp -v /var/run/docker.sock:/var/run/docker.sock -e WEAVE_PASSWORD -e WEAVE_CIDR=none weaveworks/weave:1.0.1 -iface ethwe -port 6783 -name e6:04:94:89:4a:05 -nickname ***** -iprange 172.16.0.0/8
+ CONTAINER=cddbcdbe13d794534ee15110d33ae45987e69ab50938e99d13060cd7268a3aa5
+ with_container_netns_or_die cddbcdbe13d794534ee15110d33ae45987e69ab50938e99d13060cd7268a3aa5 launch
+ with_container_netns cddbcdbe13d794534ee15110d33ae45987e69ab50938e99d13060cd7268a3aa5 launch
+ CONTAINER=cddbcdbe13d794534ee15110d33ae45987e69ab50938e99d13060cd7268a3aa5
+ docker inspect --format={{.State.Pid}} cddbcdbe13d794534ee15110d33ae45987e69ab50938e99d13060cd7268a3aa5
+ CONTAINER_PID=2.277362e+06
+ [ 2.277362e+06 = 0 ]
+ [ 2.277362e+06 = <no value> ]
+ LOCAL_IFNAME=vethwepl2.277362e+06
+ GUEST_IFNAME=vethwepg2.277362e+06
+ IP_TMPOUT=/tmp/weave_ip_out_6
+ IP_TMPERR=/tmp/weave_ip_err_6
+ rm -f /tmp/weave_ip_out_6 /tmp/weave_ip_err_6
+ STATUS=0
+ shift 1
+ launch
+ STATUS=1
+ [ ! -d /hostproc/2.277362e+06 ]
+ echo Container cddbcdbe13d794534ee15110d33ae45987e69ab50938e99d13060cd7268a3aa5 died
Container cddbcdbe13d794534ee15110d33ae45987e69ab50938e99d13060cd7268a3aa5 died
+ rm -f /tmp/weave_ip_out_6 /tmp/weave_ip_err_6
+ return 1
+ docker kill cddbcdbe13d794534ee15110d33ae45987e69ab50938e99d13060cd7268a3aa5
+ exit 1
````

Questions : How can we reset weave totally without restart of host?
One more option Observation weave bridge state is DOWN 
weave: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 65535 qdisc noqueue state DOWN
    link/ether e6:04:94:89:4a:05 brd ff:ff:ff:ff:ff:ff

Even after ip link set dev weave up does not bring this up .

**Comment by rachit1arora** [comment]
In debug mode weave launch is detecting docker version 1.3.1 but it is actually 1.6.2

docker --version
Docker version 1.6.2, build ba1f6c3/1.6.2

**Comment by bboreham** [comment]
Hi, you seem to be using quite an old version of Weave Net; is it possible for you to upgrade to the latest (1.3.1)? You can download the latest script from https://git.io/weave. (note some options have changed spelling, including `--ipalloc-range`)

Nontheless, the command `weave reset` should reset weave without restarting the host; it deletes the `weave` bridge so the next time you do `weave launch` it will be recreated.

I note that the IP range `172.16.0.0/8` is a little odd - you are asking to use the first 8 bits of the address, which is just the `172` part, but you then supply some more bits.  Did you mean `/16` ?

Weave bundled a copy of the Docker 1.3.1 client so that we could be sure of operating with older daemons, so that report is correct in context.  The latest Weave release updated this version.

The message "Usage of loopback devices is strongly discouraged for production use." is coming from Docker who warn against operating in this manner [here](https://github.com/docker/docker/blob/master/daemon/graphdriver/devmapper/README.md). But this probably doesn't relate to the problem you are seeing.

**Comment by rachit1arora** [comment]
right , we have tried weave reset  but even after that weave launch does not work . Only option working for us is host reboot . 

Migration to next version of weave is not feasible as of now  but will try it next month. 

We are debugging 2 issues 
1. Intermittently (our containers) start dying as they are not able to attach ip when we do 
docker run -e WEAVE_CIDR... 
2. When we encountered First situation what the options available to bring back the host to stable condition apart from host reboot.

**Comment by awh** [comment]
Hi @rachit1arora,

One thing I notice from the debug transcript you provided is this:

```
+ docker inspect --format={{.State.Pid}} cddbcdbe13d794534ee15110d33ae45987e69ab50938e99d13060cd7268a3aa5
+ CONTAINER_PID=2.277362e+06
```

For some reason, it looks like Docker is printing the PID in scientific notation - this is problematic because we use the PID to synthesise names for the veth pairs which are used to attach containers (including the router itself) to the weave bridge e.g.

```
+ LOCAL_IFNAME=vethwepl2.277362e+06
+ GUEST_IFNAME=vethwepg2.277362e+06
```

Those names are too long for the kernel, which imposes a fifteen character limit on interface names. This might explain why everything works for a while, and then starts to fail in a way which only a reboot fixes - if you're starting containers frequently (as you suggest you are) the PID increases to the point where Docker starts printing it in scientific notation, after which time we can no longer attach containers (including the router itself via `weave launch`) to the bridge. Rebooting the machine resets the PID counter, and everything works again for a while.

Does this sound like a possibility? 2277362 is very high for a PID - the default maximum PID on Linux is 32768:

```
$ cat /proc/sys/kernel/pid_max 
32768
```

Perhaps your distro ships with a higher value, or maybe you have modified it yourself? The maximum value for `pid_max` is 2^22 (4194304), which would be enough to allow a value of 2277362.

If you are not seeing these very high PIDs on your systems, we will need to look for another explanation as to why `docker inspect` is emitting a value of `2.277362e+06` for the container PID.

**Comment by rachit1arora** [comment]
I verified pid in our machine is this - changed to max possible 

# cat /proc/sys/kernel/pid_max
4194303

**Comment by rachit1arora** [comment]
I verified , in our distro we have changed it to max possible . 

cat /proc/sys/kernel/pid_max
4194303


In normal scenario - when every thing ok we see PID in normal range
docker inspect --format={{.State.Pid}} weave
5460

**Comment by awh** [comment]
This is a Docker bug: docker/docker#11924. Are you able to try a more recent version of Docker to see if it is fixed? In the meantime you could reduce the value of `pid_max` as a workaround.

**Comment by rachit1arora** [comment]
ok , But Is this the cause of weave launch container dying ?

**Comment by awh** [comment]
> But Is this the cause of weave launch container dying?

Let me be sure I am absolutely clear on the sequence of events:

1. Weave router container is launched, application containers are attaching successfully
2. After a while, newly launched application containers can no longer attach to the router
3. You stop and relaunch the router as an attempted remedy for the above, but now even the router won't restart
4. Reboot fixes

Is that correct? Or by "Sometime on one of the machine container start dieing when we need to apply weave ip" do you mean that the weave router container also died unexpectedly at (2)?

**Comment by awh** [comment]
If you're not adverse to patching the `weave` script yourself @bboreham suggests you could modify the line (312 in v1.0.1) that says
```
CONTAINER_PID=$(docker inspect --format='{{.State.Pid}}' $CONTAINER)
```
to read
```
CONTAINER_PID=$(docker inspect --format='{{printf "%.0f" .State.Pid}}' $CONTAINER)
```
as per the workaround outlined in [this comment](https://github.com/docker/docker/issues/11924#issuecomment-87689722). Beware this produces an error with more recent versions of Docker though!

**Comment by bboreham** [comment]
I have an alternate work-around: `-f '{{json .State.Pid}}'` seems to work in both Docker 1.6.2 and 1.9.1

**Comment by bboreham** [comment]
Sorry, didn't test that last one very well, ignore me

**Comment by awh** [comment]
@rachit1arora 

> Is that correct?

If that is correct, then yes I am fairly sure the docker bug is the cause of the weave container dying when you attempt to relaunch it. The router needs to be attached to the bridge as with any other container - the exact same code is used in the `weave` script to do it. The `Unable to find interface ethwe` message you see is the router bailing because it has not been attached to the bridge.

**Comment by rachit1arora** [comment]
@awh 
Your understanding is correct overall . Please find my response inline 


    1. Weave router container is launched, application containers are attaching successfully

--> Correct
   2.  After a while, newly launched application containers can no longer attach to the router
--> Correct and the newly launched application containers exit also. 

 3.   You stop and relaunch the router as an attempted remedy for the above, but now even the router won't restart
--> Correct
4.    Reboot fixes
--> Correct

PS : WEAVE Container DO NOT die unexpectedly . 

I will try patch on our system and will update if that helps. We do plan to move to docker 1.9 shortly but we will then move to latest version of weave also. Rather we plan on using weave driver to try out docker multihost networking . 

Question : With recent version of weave do you expect any sideeffect ? or with the latest version of weave everything will be fine ? 

On a unrelated topic . We plan on evaluating weave a other drivers with Docker 1.9 multi host networking feature. What is the recommended version of weave driver version to be used ? Will that version support fast data path ? 

Just an FYI . We plan to evaluate following drivers to compare on following points mainly  performance(on a 10Gbps network) , stability , scalability (Scale upto 50 containers spread across 50 host machines) . 

List of Network Drivers to test and compare

    Overlay Driver - Default : https://docs.docker.com/engine/userguide/networking/get-started-overlay/
    weave network plugin : https://github.com/weaveworks/docker-plugin
    gropher-net : https://github.com/gopher-net/docker-ovs-plugin
    calico driver : http://www.projectcalico.org/docker-libnetwork-is-almost-here-and-calico-is-ready/
    contiv (Cisco) : https://github.com/contiv/netplugin , multi-hos texample https://github.com/contiv/netplugin/blob/master/docs/TwoHostMultiVlanDemo.md
    MidoNet : https://blog.midonet.org/kuryr-midonet-networking-docker-1-9/

If you guys have done any comparison study , please share with us . 
Any comments on this are welcome . 

Also Thanks for the excellent support provided on various issues and help.

**Comment by rade** [comment]
> With recent version of weave do you expect any sideeffect ? or with the latest version of weave everything will be fine?

I am not quite sure what you are asking here. As per our docs, weave works fine with all recent Dockers, including 1.9. And the issue you ran into was fixed in Docker 1.7.

> On a unrelated topic

In future, please ask on [weave-users](https://groups.google.com/a/weave.works/forum/#!forum/weave-users)

> We plan on evaluating weave a other drivers with Docker 1.9 multi host networking feature. What is the recommended version of weave driver version to be used ?

From [the Weave Net 1.3.0 release notes](https://github.com/weaveworks/weave/releases/tag/v1.3.0):

> This release includes a Docker Plugin

i.e. the Weave Docker Network Driver Plugin is part of Weave Net 1.3.0 and later. For more details see [the docs](http://docs.weave.works/weave/latest_release/features.html#plugin).

> Will that version support fast data path ?

Yes.
